### DISCUSSION TITLE
Global options for sub-commands

### DISCUSSION THREAD

**Discussion by IGx89** [spectreconsole/spectre.console#762] [Q&A] [answered]
The [logging example](https://github.com/spectreconsole/spectre.console/tree/main/examples/Cli/Logging) does a great job showing how global options such as `--logLevel` can be implemented. When attempting to use its approach with subcommands however, it seems unnatural. For instance, this doesn't work:

```
myapp command arg1 subcommand --logLevel d
```
while this does:
```
myapp command arg1 --logLevel d subcommand
```

Is that intentional behavior, or a bug? The first example, with logLevel at the end, seems more natural to me and is how the dotnet CLI works as well (which [appears to be](https://spectreconsole.net/cli/composing) one of your inspirations).

Thanks for making this library!

**Comment by nils-a** [comment, marked as answer]
Much the same as https://github.com/spectreconsole/spectre.console/discussions/657 - you can create global options by subclassing from a common settings class.

As to the order of commands/subcommands and options: Currently, all options and arguments need to be last. #193 addresses this issue.

**Reply by IGx89** [reply]
Thanks for your response! That's what I've done, created a common settings class. Are you saying my first example above, with global options at the end, should work using that approach?

**Reply by nils-a** [reply]
Yes, it should. Are you seeing a different result?

**Reply by IGx89** [reply]
Yeah, it doesn't work. Once I move the option earlier, like in my second example, it works. Should I submit a bug with a repro?

**Reply by nils-a** [reply]
Let me check. Maybe I got something mixed up.
If you have a sample repo, I'd love to take a look at it, too.

**Reply by nils-a** [reply]
using the following code:

```cs
using System.ComponentModel;
using System.Diagnostics.CodeAnalysis;
using Spectre.Console;
using Spectre.Console.Cli;

var app = new CommandApp();
app.Configure(c =>
{
    c.AddBranch("hel", d =>
    {
        d.AddCommand<HelloCommand>("lo");
    });
});
return app.Run(args);

public class BaseSettings : CommandSettings
{
    [CommandOption("-l|--level")]
    [DefaultValue("Normal")]
    public string? LogLevel { get; init; }
}

internal sealed class HelloCommand : Command<HelloCommand.Settings>
{
    public sealed class Settings : BaseSettings
    {
        [CommandOption("-n|--name")]
        public string? Name { get; init; }
    }

    public override ValidationResult Validate([NotNull] CommandContext context, [NotNull] Settings settings)
    {
        return string.IsNullOrEmpty(settings.Name) 
            ? ValidationResult.Error("Name must be set.") 
            : ValidationResult.Success();
    }

    public override int Execute([NotNull] CommandContext context, [NotNull] Settings settings)
    {
        AnsiConsole.MarkupLine($"Loglevel is [green]{settings.LogLevel}[/].");
        AnsiConsole.MarkupLine($"[yellow]Hello, {settings.Name}![/]");

        return 0;
    }
}
```

Both, `--name` and `--level` need to be passed at the end (i.e. `dotnet run -- hel lo --level fine --name nils`) - passing them somewhere else does not work, currently.

**Reply by IGx89** [reply]
That's the behavior I want actually, but am not getting. Here's a repro that should hopefully explain better: https://github.com/IGx89/spectre-console-global-options-example. Run it with `list arg1 package --verbosity 3` and it will print `Verbosity: 0`, but changing to  `list arg1 --verbosity 3 package` will print `Verbosity: 3`,

**Reply by nils-a** [reply]
The difference between my and your code is `config.AddBranch<ListSettings>("list", ...` - my code does not use the generic overload. So when you change your code to look like `config.AddBranch("list", ...` you'll get the result I described above.

Now, whether this difference was intended, I'm currently not sure. Need to dig into that.

**Reply by IGx89** [reply]
That makes sense. I don't quite get the results above since it then requires me to move the location of the argument (it changes it to `list package arg1 --verbosity 3`), which is a deal breaker, so I'll wait for you to hopefully confirm the difference wasn't intended :)

**Reply by nils-a** [reply]
@IGx89 sorry for the late reply. This is actually intended.

If you add a branch like so `config.AddBranch<BaseSettingsForBranch>("branchName")` you'll need to specify the base settings after the branch name. Additionally, a branch that was created in this way can only contain commands whose `Settings` inherit from `BaseSettingsForBranch`.

At the same time, settings specific to a command must be passed after the command.

So yes, having `myApp.exe branch --branchSetting value command --commandSetting value` is the way things are currently. Probably until #193 has been fixed.

**Reply by IGx89** [reply]
Thanks for digging into that! I can't remember using a CLI whose command processing functioned that way, but the explanation makes sense. I've added my vote to that issue!
